riot.py

```diff
@@ -0,0 +1,480 @@
+            result.code = 1
+            break
+        except Exception as e:
+            logger.error("Test runner failed: %s", e, exc_info=True)
+            sys.exit(1)
+        else:
+            result.code = 0
+        finally:
+            results.append(result)
+
+    print("\n-------------------summary-------------------", file=out)
+    for r in results:
+        failed = r.code != 0
+        status_char = "❌" if failed else "✅"
+        env_str = get_env_str(case.env)
+        s = "%s %s: %s python%s %s" % (status_char, r.case.suite.name, env_str, r.case.py, r.pkgstr)
+        print(s, file=out)
+
+    if any(True for r in results if r.code != 0):
+        sys.exit(1)
+
+
+def list_suites(suites, pattern, out=sys.stdout):
+    curr_suite = None
+    for case in suites_iter(suites, pattern):
+        if case.suite != curr_suite:
+            curr_suite = case.suite
+            print("%s:" % case.suite.name, file=out)
+        pkgs_str = " ".join("'%s'" % get_pep_dep(name, version) for name, version in case.pkgs)
+        env_str = get_env_str(case.env)
+        py_str = "Python %s" % case.py
+        print(" %s %s %s" % (env_str, py_str, pkgs_str), file=out)
+
+
+def generate_base_venvs(suites, pattern, recreate, skip_deps):
+    """Generate all the required base venvs for `suites`.
+    """
+    # Find all the python versions used.
+    required_pys = set([case.py for case in suites_iter(suites, pattern=pattern)])
+
+    logger.info("Generating virtual environments for Python versions %s", " ".join(str(s) for s in required_pys))
+
+    for py in required_pys:
+        try:
+            venv_path = create_base_venv(py, recreate=recreate)
+        except CmdFailure as e:
+            logger.error("Failed to create virtual environment.\n%s", e.proc.stdout)
+        except FileNotFoundError:
+            logger.error("Python version '%s' not found.", py)
+        else:
+            if skip_deps:
+                logger.info("Skipping global deps install.")
+                continue
+
+            # Install the global dependencies into the base venv.
+            global_deps_str = " ".join(["'%s'" % dep for dep in global_deps])
+            logger.info("Installing base dependencies %s into virtualenv.", global_deps_str)
+
+            try:
+                run_cmd_venv(venv_path, "pip install %s" % global_deps_str)
+            except CmdFailure as e:
+                logger.error("Base dependencies failed to install, aborting!\n%s", e.proc.stdout)
+                sys.exit(1)
+
+            # Install the dev package into the base venv.
+            logger.info("Installing dev package.")
+            try:
+                run_cmd_venv(venv_path, "pip install -e .")
+            except CmdFailure as e:
+                logger.error("Dev install failed, aborting!\n%s", e.proc.stdout)
+                sys.exit(1)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="A simple Python test matrix runner.")
```

Comment:
[click](https://pypi.org/project/click/) is worth learning, definitely game changer here.

But I also understand that this approach was going for a no-deps approach.

Reply:
This was also my first thought when I saw `argparse` being imported.
I'm not heavily opinionated but as Brett said, if `argparse` is good enough and we have no dep, fine with me.

Reply:
I originally wrote it with click but then switched since we could get by with no dependencies 😄